_posts/2020-10-09-from-sprockets-to-webpacker.markdown

```diff
@@ -0,0 +1,315 @@
+$ yarn add @rails/ujs
+```
+
+```js
+// app/javascript/packs/application.js
+require('@rails/ujs').start()
+```
+
+Each 3rd party package that you use can have its own way to use it with Webpack.
+
+### Moving The Styleguide
+
+Now we need to do the biggest change: replace the `fastruby-styleguide` Ruby gem with the new `fastruby-io-styleguide` node module.
+
+First we removed the gem from our gemfile and added the node module using yarn directly from github running `yarn add "fastruby/styleguide#gh-pages"` (`#...` is the branch name we want).
+
+Now we have to change how we import the assets in both the SCSS and the JS files:
+
+```js
+// app/assets/stylesheets/application.js
+// before
+@import "fastruby/styleguide";
+
+// after
+@import "fastruby-io-styleguide";
+```
+
+```js
+// app/javascript/packs/application.js
+// added at the top
+import "fastruby-io-styleguide"
+```
+
+And we can finally clear the previous `application.js` at `app/assets/javascript/application.js`.
+
+### jQuery Integration
+
+This last change created a few issues related to jQuery: the `jQuery` object was not available globally (so Bootstrap failed), and jQuery's `ready` callback was not triggered properly.
+
+To fix the `undefined jQuery` error, we need to configure Webpack to expose it using this config at `config/webpack/environment.js`:
+
+```js
+const { environment } = require('@rails/webpacker')
+
+const webpack = require('webpack')
+environment.plugins.prepend(
+  'Provide',
+  new webpack.ProvidePlugin({
+    $: 'jquery/src/jquery',
+    jQuery: 'jquery/src/jquery'
+  })
+)
+
+module.exports = environment
+```
+
+To fix the issue with jQuery's `ready` callback not triggered, we fixed it by just replacing all `$(document).on('ready', ...)` calls (and similar ones) to `document.addEventListener('DOMContentLoaded', ...)`.
+
+Now all our tests are green.
+
+### Remove `app/assets/javascript`
+
+Now that our original `application.js` file is empty, we can remove it, but we need to fix Sprockets' config at `app/assets/config/manifest.js`. Remove the line: `//= link_directory ../javascript .js` so it doesn't search for a javascript directory that doesn't exist anymore.
+
+Finally, we can remove the `<script>` tag in the head of our layout by removing the old `javascript_include_tag 'application'` call.
+
+## Configure CircleCI
+
+Since we are using CircleCI as a continuos integration service an we were not using Yarn to handle 3rd party node modules before, we need to fix the configuration so CircleCI installs all Yarn packages before running the tests.
+
+Look for your `bundle install` call in the `.circle/config.yml` config file and add the `yarn install` command after it.
+
+## Conclusion
+
+It takes some time and a bit of a mindset change since Sprockets and Webpack works differently, but now we are using the current Rails standard to handle assets. Thanks to this and Webpack's popularity, we can now use modern JS features, frameworks, and tools more easily.
```

Comment:
```suggestion
It takes some time and a bit of a mindset change since Sprockets and Webpack work differently, but now we are using the current Rails standard to handle assets. Thanks to this and Webpack's popularity, we can now use modern JS features, frameworks, and tools more easily.
```